Core.UnitTests/CodeGeneration/ReflectionEmit/MemberEmitterTest.cs

```diff
@@ -102,40 +105,48 @@ public void AddConstructor ()
       var expectedParameterTypes = new[] { typeof (string), typeof (int).MakeByRefType() };
       var expectedCallingConventions = CallingConventions.Standard | CallingConventions.HasThis;
 
-      var constructorBuilderMock = MockRepository.GenerateStrictMock<IConstructorBuilder>();
+      var constructorBuilderMock = new Mock<IConstructorBuilder> (MockBehavior.Strict);
       _typeBuilderMock
-          .Expect (mock => mock.DefineConstructor (expectedAttributes, expectedCallingConventions, expectedParameterTypes))
-          .Return (constructorBuilderMock);
-      constructorBuilderMock.Expect (mock => mock.RegisterWith (_emittableOperandProviderMock, constructor));
-
-      SetupDefineCustomAttribute (constructorBuilderMock, constructor);
-      var parameterBuilderMock = SetupDefineParameter (constructorBuilderMock, 1, "p1", ParameterAttributes.In);
-      SetupDefineCustomAttribute (parameterBuilderMock, constructor.MutableParameters[0]);
-      SetupDefineParameter (constructorBuilderMock, 2, "p2", ParameterAttributes.Out);
+          .Setup (mock => mock.DefineConstructor (expectedAttributes, expectedCallingConventions, expectedParameterTypes))
+          .Returns (constructorBuilderMock.Object)
+          .Verifiable();
+      constructorBuilderMock.Setup (mock => mock.RegisterWith (_emittableOperandProviderMock.Object, constructor)).Verifiable();
+
+      var customAttributeTargetBuilderMock = constructorBuilderMock.As<ICustomAttributeTargetBuilder>();
```

Comment:
If I understood your question correctly: Moq's `As<T>` doesn't return an actual cast object but a new mock that can (or rather has to) be setup and verified separately

Reply:
okay but why do this? It also sounds wrong to me to split the instances? Unless there's a reason to change this (i.e. it no longer works the old way) we shouldn't make semantic changes to the tests.

Reply:
`Mock<T>` doesn't seem to be covariant, in order to use the custom Setup methods we already had we needed to go one step up the inheritance hierarchy. We can start thinking about adjusting the setup method signatures or the mock types but imo this was the change that preserved the test semantics the most

Reply:
I see. Still don't like it. Maybe, when you know the syntax, it's perfectly obvious, but having multiple instaces for the same type just seems weird. can't we make the Helper-Setup-Methods generic with the respecive type constraints and type inference? that should take care of everything while only chaning the line where the helper-method is declared.

Reply:
I found out that when calling `.As<T>` setups are performed on the object returned by it AND on the original object, so yes, it looks like we can leave the tests and "generify" the custom setup methods

Reply:
ah okay, that's good. I'm guessing setup and verify are performed. still prefer the genric version when if the compiler agrees because it's just more explicit what's happening and why (i.e. you don't need to understand moq to understand that all is OK)